### PR TITLE
Implement batch snowball

### PR DESCRIPTION
**Pull request by vadmeste**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
Enable snowball by default with object sizes < 5 MiB when replicating objects from a local source

Example of replicate job addition:

````
replicate:
  apiVersion: v1
  # source of the objects to be replicated
  source:
    type: minio # valid values are "s3" or "minio"
    bucket: testbucket
    snowball:
      disable: false # optionally turn-off snowball archive transfer
      batch: 100 # upto this many objects per archive
      inmemory: true # indicates if the archive must be staged locally or in-memory
      compress: false # S2/Snappy compressed archive
      smallerThan: 5MiB # create archive for all objects smaller than 5MiB
      skipErrs: false # skips any source side read() errors

  # target where the objects must be replicated
  target:
    type: minio
    bucket: testbucket
    endpoint: "http://localhost:9002"
    credentials:
      accessKey: minioadmin
      secretKey: minioadmin
```

## Motivation and Context


## How to test this PR?


## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by vadmeste** [comment]
@harshavardhana this is an example of yaml file with replication snowball feature; (under flags subsection)

```
replicate:
  apiVersion: v1
  # source of the objects to be replicated
  source:
    type: minio # valid values are "s3" or "minio"
    bucket: testbucket

  # target where the objects must be replicated
  target:
    type: minio
    bucket: testbucket
    endpoint: "http://localhost:9002"
    credentials:
      accessKey: minioadmin
      secretKey: minioadmin

  flags:
    snowball:
      disable: false # optionally turn-off snowball archive transfer
      batch: 100 # upto this many objects per archive
      inmemory: true # indicates if the archive must be staged locally or in-memory
      compress: false # S2/Snappy compressed archive
      smallerThan: 5MiB # create archive for all objects smaller than 5MiB
      skipErrs: false # skips any source side read() errors
```

Snowball will be automatically enabled if the source is local and the target is remote, right ?

File: go.mod [hunk 1]
```diff
@@ -50,7 +50,7 @@ require (
 	github.com/minio/highwayhash v1.0.2
 	github.com/minio/kes-go v0.2.0
 	github.com/minio/madmin-go/v3 v3.0.29
-	github.com/minio/minio-go/v7 v7.0.64-0.20231119012610-6eebdd6d5eba
+	github.com/minio/minio-go/v7 v7.0.64-0.20231120033707-0108287f5fc3
```
Comment (harshavardhana) [inline go.mod]: A new release is done you can use that.

File: cmd/batch-job-common-types.go [hunk 2]
```diff
@@ -81,3 +83,34 @@ func (r BatchJobRetry) Validate() error {
...
+// Validate the snowball parameters in the job description
+func (b BatchJobSnowball) Validate() error {
+	if *b.Batch <= 0 {
+		return errors.New("batch number should be non positive zero")
+	}
+	_, err := humanize.ParseBytes(*b.SmallerThan)
+	return err
+}
+
+// TODO: add validation
```
Comment (harshavardhana) [inline cmd/batch-job-common-types.go]:
```suggestion
// TODO: add validation
```

??

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/utils.go [hunk 3]
```diff
@@ -1265,3 +1265,15 @@ func stringsHasPrefixFold(s, prefix string) bool {
 	// Test match with case first.
 	return len(s) >= len(prefix) && (s[0:len(prefix)] == prefix || strings.EqualFold(s[0:len(prefix)], prefix))
 }
+
+func boolPtr(v bool) *bool {
```
Comment (jiuker) [inline cmd/utils.go]:
```golang
func Ptr[T any](a T) *T {
	return &a
}
```
Better?

**Review by jiuker — APPROVED** [review]

File: cmd/batch-handlers.go [hunk 4]
```diff
@@ -941,8 +996,71 @@ func (r *BatchJobReplicateV1) Start(ctx context.Context, api ObjectLayer, job Ba
 	if err != nil {
 		return err
 	}
+
 	c.SetAppInfo("minio-"+batchJobPrefix, r.APIVersion+" "+job.ID)
 
+	var (
+		walkCh = make(chan ObjectInfo, 100)
```
Comment (jiuker) [inline cmd/batch-handlers.go:1003]: max(r.Source.Snowball.Batch,100) should be fine？

**Review by jiuker — CHANGES_REQUESTED** [review]

File: cmd/batch-handlers.go [hunk 5]
```diff
@@ -941,8 +996,71 @@ func (r *BatchJobReplicateV1) Start(ctx context.Context, api ObjectLayer, job Ba
...
+				if len(batch) > 0 {
+					if err := r.writeAsArchive(ctx, api, cl, batch); err != nil {
+						logger.LogIf(ctx, err)
+						for _, b := range batch {
+							slowCh <- b
+						}
+					}
+					batch = batch[:0]
+				}
+			}
```
Comment (jiuker) [inline cmd/batch-handlers.go:1060]: Should close `slowCh`  after the `walkCh` is closed. Or  reading from `slowCh` will never be end.

**Review by jiuker — CHANGES_REQUESTED** [review]

File: cmd/utils.go (hunk 3 above)
Comment (vadmeste) [reply cmd/utils.go]: yes!

File: cmd/batch-handlers.go (hunk 5 above)
Comment (vadmeste) [reply cmd/batch-handlers.go:1060]: Good catch!

**Review by harshavardhana — APPROVED** [review]

**Review by jiuker — APPROVED** [review]
LGTM

**Comment by harshavardhana** [comment]
@vadmeste please fix the go.mod

**Pull request merged**